### PR TITLE
MGMT-2086: upgrade to GORM v2

### PR DESCRIPTION
**Pull request by ori-amizur**
# Assisted Pull Request

## Description
    - Upgrade to gorm v1.22.3. Gorm is now in gorm.io/gorm.  GORM v1 will
      eventually not be supported.
    - Postgres driver is in its own repo gorm.io/driver/postgres v1.2.1
    - gormigrate has been updated to v2.0.0.
    - gorm has some changes
      * UpdatedAt, CreatedAt have to be time.Time to
        be updated by gorm.  Making use of go-swagger x-go-type.
      * DeletedAt field has to be a special field of type gorm.DeletedAt.
        Making use of go-swagger x-go-type.
      * Retruned dates from gorm might not be binary equal to the value that
        was inserted to the database, but time.Time.Equal function is used
        to verify the equality
      * Some chained methods of gorm.DB may not be reused.  For example
        the value of db.Unscoped() cannot be used more than once.  Each such
        invocation needs to call Unscoped() again.  The known chained
        methods are: Unscoped, Where, Limit, Order, Clauses
        (for "FOR UPDATE" query)
      * GORM has Migrator() function that contains implementation of new
        migration API.
      * gorm.DB cannot be closed directly, but the underlying sqlDB needs to
        be closed.
      * Global updates are not allowed unless specified directly.
      * Auto increment ID is now bigint instead of integer.  Therefore,
        events table had to be migrated.

## List all the issues related to this PR

- [ ] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [x] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Review by nmagnezi — CHANGES_REQUESTED** [review]
could you please add some details (commit-msg) about the motivation for this upgrade?

**Comment by ori-amizur** [comment]
> could you please add some details (commit-msg) about the motivation for this upgrade?

End of life for form v1

**Comment by filanov** [comment]
/hold 
@ori-amizur did @pawanpinjarkar upgrade test passed with this PR?

**Comment by ori-amizur** [comment]
> /hold @ori-amizur did @pawanpinjarkar upgrade test passed with this PR?

Not yet

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -4225,7 +4214,6 @@ func (b *bareMetalInventory) DisableHost(ctx context.Context, params installer.D
 
 	txSuccess := false
 	tx := b.db.Begin()
-	tx = transaction.AddForUpdateQueryOption(tx)
```
Comment (filanov) [inline internal/bminventory/inventory.go:4228]: What is the downside of doing it that way?
I think that other developers will forget to put this no?

Comment (ori-amizur) [reply internal/bminventory/inventory.go:4228]: I think we need to modify the common query functions that they will support this option.
The "for update" clause has to issued on the initial query in a transaction.  After that, the relevant 
records are locked until the end of the transaction

Comment (ori-amizur) [reply internal/bminventory/inventory.go:4228]: But this will be probably part of a different task

File: internal/cluster/cluster.go [hunk 2]
```diff
@@ -1101,22 +1102,23 @@ func (m Manager) PermanentClustersDeletion(ctx context.Context, olderThan strfmt
...
 			&models.Event{},
 			&models.MonitoredOperator{},
 			&models.ClusterNetwork{},
 			&models.ServiceNetwork{},
 			&models.MachineNetwork{},
-		}); err != nil {
-			m.log.WithError(err).Warnf("Failed deleting cluster records from db for cluster %s", c.ID.String())
+		}
+		for _, model := range modelsToDelete {
+			if err := common.DeleteRecordsByClusterID(m.db.Unscoped(), *c.ID, []interface{}{model}); err != nil {
```
Comment (filanov) [inline internal/cluster/cluster.go:1113]: why is unscoped used here? shouldn't our GC handle that? 
We want to keep all the elements in order to investigate them in case of errors

File: internal/host/hostutil/update_host.go [hunk 3]
```diff
@@ -103,7 +103,13 @@ func UpdateHost(_ logrus.FieldLogger, db *gorm.DB, infraEnvId strfmt.UUID, hostI
 	}
 
 	if dbReply.RowsAffected == 0 && !hostExistsInDB(db, hostId, infraEnvId, updates) {
-		return nil, errors.Errorf("failed to update host %s from infra-env %s. nothing has changed", hostId, infraEnvId)
+		host, err := common.GetHostFromDB(db, infraEnvId.String(), hostId.String())
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to update host %s from infra-env %s. nothing has changed", hostId, infraEnvId)
+		} else {
+			return nil, errors.Errorf("failed to update host %s from infra-env %s status %s expected status %s but host exists in db", hostId, infraEnvId, swag.StringValue(host.Status),
```
Comment (filanov) [inline internal/host/hostutil/update_host.go:110]: please break long lines.

File: internal/host/transition.go [hunk 4]
```diff
@@ -37,7 +37,7 @@ var resetFields = append(resetProgressFields, "inventory", "", "bootstrap", fals
 var restFieldsOnUnbind = append(append(resetProgressFields, resetLogsField...), "cluster_id", nil, "kind", swag.String(models.HostKindHost), "connectivity", "", "domain_name_resolutions", "",
 	"free_addresses", "", "images_status", "", "installation_disk_id", "", "installation_disk_path", "", "machine_config_pool_name", "",
 	"role", "auto-assign", "api_vip_connectivity", "", "suggested_role", "",
-	"progress_stages", nil, "stage_started_at", strfmt.DateTime(time.Time{}), "stage_updated_at", strfmt.DateTime(time.Time{}))
```
Comment (filanov) [inline internal/host/transition.go:40]: why was that removed?

File: internal/host/transition.go [hunk 5]
```diff
@@ -259,7 +259,7 @@ func (th *transitionHandler) PostResetHost(sw stateswitch.StateSwitch, args stat
 		return errors.New("PostResetHost invalid argument")
 	}
 
-	extra := append(append(make([]interface{}, 0), "StatusUpdatedAt", strfmt.DateTime(time.Now())), resetLogsField...)
```
Comment (filanov) [inline internal/host/transition.go:262]: why was that removed?

File: internal/host/transition.go [hunk 6]
```diff
@@ -439,7 +439,7 @@ func (th *transitionHandler) PostResettingPendingUserAction(sw stateswitch.State
 	}
 
 	return th.updateTransitionHost(params.ctx, logutil.FromContext(params.ctx, th.log), params.db, sHost,
-		statusInfoResettingPendingUserAction, "StatusUpdatedAt", strfmt.DateTime(time.Now()))
```
Comment (filanov) [inline internal/host/transition.go:442]: why was that removed?

**Comment by filanov** [comment]
@ori-amizur lets run a swarm test on that as well @omertuc can probably help with that

File: internal/cluster/cluster.go (hunk 2 above)
Comment (ori-amizur) [reply internal/cluster/cluster.go:1113]: This is a function for permanent deletion

File: internal/host/transition.go (hunk 4 above)
Comment (ori-amizur) [reply internal/host/transition.go:40]: The field doesn't exist

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (filanov) [reply internal/bminventory/inventory.go:4228]: Yes it is out of scope for this PR, please open another ticket for that.

File: internal/cluster/cluster.go (hunk 2 above)
Comment (filanov) [reply internal/cluster/cluster.go:1113]: right, missed that.

File: internal/host/transition.go (hunk 5 above)
Comment (ori-amizur) [reply internal/host/transition.go:262]: If we change status, this field is set here https://github.com/openshift/assisted-service/blob/3c6c20c6f897a2a0378a6ff41c141028c52b8a9b/internal/host/hostutil/update_host.go#L61
So there is an attempt to change this field twice which causes an error

File: internal/host/transition.go (hunk 6 above)
Comment (ori-amizur) [reply internal/host/transition.go:442]: Same as above

File: internal/host/hostutil/update_host.go (hunk 3 above)
Comment (ori-amizur) [reply internal/host/hostutil/update_host.go:110]: OK

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (ori-amizur) [reply internal/bminventory/inventory.go:4228]: https://issues.redhat.com/browse/MGMT-8252

File: internal/migrations/20201202140700_change_image_ssh_key_text_test.go [hunk 7]
```diff
@@ -45,26 +47,28 @@ var _ = Describe("changeOverridesToText", func() {
 	})
 
 	It("Migrates down and up", func() {
-		t, err := getColumnType(db, &common.Cluster{}, "image_ssh_public_key")
-		Expect(err).ToNot(HaveOccurred())
-		Expect(t).To(Equal("TEXT"))
+		if db.Migrator().HasColumn(&common.Cluster{}, "image_ssh_public_key") {
```
Comment (carbonin) [inline internal/migrations/20201202140700_change_image_ssh_key_text_test.go:50]: Why add this? Pretty sure the column still exists.

File: internal/migrations/20210713123129_populate_infra_env_test.go [hunk 8]
```diff
@@ -50,8 +50,7 @@ var _ = Describe("populateInfraEnv", func() {
 			ID: &hostID,
 			// [TODO] - currently we check migration by adding dummy value and seeing it replaced
 			// in case we add host v1 model to the swagger, then we can replace the models.Host with it
-			InfraEnvID: strfmt.UUID(uuid.New().String()),
-			ClusterID:  &clusterID,
+			ClusterID: &clusterID,
```
Comment (carbonin) [inline internal/migrations/20210713123129_populate_infra_env_test.go:53]: Why is this changing? What does this have to do with gorm v2?

File: internal/migrations/migrations.go [hunk 9]
```diff
@@ -14,6 +14,7 @@ func MigratePre(db *gorm.DB) error {
 func pre() []*gormigrate.Migration {
 	preMigrations := []*gormigrate.Migration{
 		populateInfraEnv(),
+		modifyEventsId(),
```
Comment (carbonin) [inline internal/migrations/migrations.go:17]: Don't put things in preMigrations unless there is a _very_ good reason. Ideally this wouldn't exist, so if you do need to put this migration here, please add a full explanation to the commit message and PR description.

See: https://github.com/openshift/assisted-service/blob/master/docs/dev/migrations.md#pre-migrations

Comment (ori-amizur) [reply internal/migrations/migrations.go:17]: The id has to be migrated before AutoMigrate is being run.  Otherwise, AutoMigrate fails

File: internal/bminventory/inventory.go [hunk 10]
```diff
@@ -1708,9 +1708,6 @@ func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params
 
 	// prepare cluster and hosts for installation
 	err = b.db.Transaction(func(tx *gorm.DB) error {
-		// in case host monitor already updated the state we need to use FOR UPDATE option
-		tx = transaction.AddForUpdateQueryOption(tx)
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:1712]: FOR UPDATE is for queries only.  We are not doing it here

File: internal/bminventory/inventory.go [hunk 11]
```diff
@@ -4238,7 +4226,7 @@ func (b *bareMetalInventory) DisableHost(ctx context.Context, params installer.D
 		}
 	}()
 
-	host, err := common.GetHostFromDB(b.db, params.ClusterID.String(), params.HostID.String())
+	host, err := common.GetHostFromDB(transaction.AddForUpdateQueryOption(tx), params.ClusterID.String(), params.HostID.String())
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:4229]: Maybe.  We will do a refactor concerning these functions. We should not have a special function "ForUpdate", Instead these functions should receive options, and one of them should be "ForUpdate".  Will be done in a different task.

File: internal/bminventory/inventory.go [hunk 12]
```diff
@@ -6791,7 +6776,6 @@ func (b *bareMetalInventory) V2UpdateHostInternal(ctx context.Context, params in
 
 	txSuccess := false
 	tx := b.db.Begin()
-	tx = transaction.AddForUpdateQueryOption(tx)
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:6794]: If it is not part of the transaction, then it has no affect.  So it means it is not needed (unless there is a known bug on it).

File: internal/cluster/cluster.go [hunk 13]
```diff
@@ -481,14 +481,17 @@ func (m *Manager) SkipMonitoring(c *common.Cluster) bool {
...
-		dbWithCondition := common.LoadTableFromDB(m.db, common.HostsTable, "status <> ?", models.HostStatusDisabled)
-		dbWithCondition = common.LoadClusterTablesFromDB(dbWithCondition, common.HostsTable)
-		dbWithCondition = dbWithCondition.Where("status NOT IN (?)", noNeedToMonitorInStates)
-		m.monitorQueryGenerator = common.NewMonitorQueryGenerator(m.db, dbWithCondition, m.MonitorBatchSize)
+			dbWithCondition := common.LoadTableFromDB(db, common.HostsTable, "status <> ?", models.HostStatusDisabled)
+			dbWithCondition = common.LoadClusterTablesFromDB(dbWithCondition, common.HostsTable)
+			dbWithCondition = dbWithCondition.Where("status NOT IN (?)", noNeedToMonitorInStates)
+			return dbWithCondition
+		}
+		m.monitorQueryGenerator = common.NewMonitorQueryGenerator(m.db, prepareDB, m.MonitorBatchSize)
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/cluster/cluster.go:494]: As I wrote in the PR 
**Some chained methods of gorm.DB may not be reused.  For example
the value of db.Unscoped() cannot be used more than once.  Each such
invocation needs to call Unscoped() again.  The known chained
methods are: Unscoped, Where, Limit, Order, Clauses**

In this case we have a reused where clause.  As of now, it cannot be used this way in GORM v2.  So I had to change it.
Please see the issue in gorm https://github.com/go-gorm/gorm/issues/4847

File: internal/cluster/common.go [hunk 14]
```diff
@@ -114,7 +114,7 @@ func clusterExistsInDB(db *gorm.DB, clusterId strfmt.UUID, where map[string]inte
 	return db.Select("id").Take(&cluster, where).Error == nil
 }
 
-func UpdateCluster(log logrus.FieldLogger, db *gorm.DB, clusterId strfmt.UUID, srcStatus string, extra ...interface{}) (*common.Cluster, error) {
+func UpdateCluster(_ logrus.FieldLogger, db *gorm.DB, clusterId strfmt.UUID, srcStatus string, extra ...interface{}) (*common.Cluster, error) {
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/cluster/common.go:117]: Can keep.

File: internal/cluster/common.go [hunk 15]
```diff
@@ -133,9 +133,12 @@ func UpdateCluster(log logrus.FieldLogger, db *gorm.DB, clusterId strfmt.UUID, s
 	}
 
 	if dbReply.RowsAffected == 0 && !clusterExistsInDB(db, clusterId, updates) {
-		return nil, errors.Errorf("failed to update cluster %s. nothing has changed", clusterId)
+		cluster, err := common.GetClusterFromDB(db, clusterId, common.SkipEagerLoading)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to update cluster %s. nothing has changed", clusterId)
+		}
+		return nil, errors.Errorf("failed to update cluster %s but it exists in db status %s expected status %s", clusterId, swag.StringValue(cluster.Status), srcStatus)
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/cluster/common.go:140]: Helped me to debug things.  Thought it will be easier for others too.

File: internal/cluster/registrar.go [hunk 16]
```diff
@@ -50,14 +50,18 @@ func (r *registrar) registerCluster(ctx context.Context, cluster *common.Cluster
...
 	}
 
-	queryParams := []string{"id = ?", cluster.ID.String()}
-	if err := tx.First(&cluster, queryParams).Error; err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+	queryParams := []interface{}{"id = ?", cluster.ID.String()}
+	var (
+		queriedCluster common.Cluster
+		err            error
+	)
+	if err = tx.First(&queriedCluster, queryParams...).Error; err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/cluster/registrar.go:58]: Can change

File: internal/cluster/registrar.go [hunk 17]
```diff
@@ -66,24 +70,20 @@ func (r *registrar) registerCluster(ctx context.Context, cluster *common.Cluster
 		}
 	}
 
-	for _, tableName := range common.ClusterSubTables {
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/cluster/registrar.go:69]: Right

File: internal/common/db.go [hunk 18]
```diff
@@ -137,8 +138,8 @@ const (
 var ClusterSubTables = [...]string{HostsTable, MonitoredOperatorsTable, ClusterNetworksTable, ServiceNetworksTable, MachineNetworksTable}
 
 func AutoMigrate(db *gorm.DB) error {
-	return db.AutoMigrate(&models.MonitoredOperator{}, &Host{}, &Cluster{}, &Event{}, &InfraEnv{},
-		&models.ClusterNetwork{}, &models.ServiceNetwork{}, &models.MachineNetwork{}).Error
+	return db.AutoMigrate(&models.MonitoredOperator{}, &InfraEnv{}, &Host{}, &Cluster{}, &Event{},
+		&models.ClusterNetwork{}, &models.ServiceNetwork{}, &models.MachineNetwork{})
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/common/db.go:142]: Will change back

File: internal/common/monitor_query_generator.go [hunk 19]
```diff
@@ -123,21 +124,21 @@ func (t *timedQuery) Next() ([]*Cluster, error) {
...
-	batchSize       int
+	lastInvokeTime time.Time
+	calls          int64
+	db             *gorm.DB
+	prepareDB      func(*gorm.DB) *gorm.DB
+	batchSize      int
 }
 
-func NewMonitorQueryGenerator(db, dbWithCondition *gorm.DB, batchSize int) *MonitorClusterQueryGenerator {
+func NewMonitorQueryGenerator(db *gorm.DB, prepareDB func(*gorm.DB) *gorm.DB, batchSize int) *MonitorClusterQueryGenerator {
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/common/monitor_query_generator.go:134]: OK

File: internal/host/hostcommands/install_cmd_test.go [hunk 20]
```diff
@@ -460,7 +460,7 @@ var _ = Describe("installcmd arguments", func() {
 			verifyArgInCommand(stepReply[0].Args[1], "--installer-args", fmt.Sprintf("'%s'", `["--append-karg","nameserver=8.8.8.8","-n","--copy-network"]`), 1)
 		})
 		It("non-empty installer args with copy network with static ip config", func() {
-			db.Model(&cluster).Update("image_static_ips_config", "rkhkjgdfd")
+			db.Model(&cluster).Update("image_static_network_config", "rkhkjgdfd")
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/host/hostcommands/install_cmd_test.go:463]: Column image_static_ips_config does not exist

File: internal/migrations/20201202140700_change_image_ssh_key_text_test.go (hunk 7 above)
Comment (ori-amizur) [reply internal/migrations/20201202140700_change_image_ssh_key_text_test.go:50]: Will remove

File: internal/bminventory/inventory.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/bminventory/inventory.go:1712]: `setBootstrapHost` (more specifically `getKnownMastersNodesIds`)  queries the hosts from the database then updates one to be the bootstrap.

File: internal/bminventory/inventory.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/bminventory/inventory.go:4229]: > Currently, we need the function GetClusterFromDBForUpdate because it adds "ForUpdate" to the preloads

:+1: makes sense

File: internal/bminventory/inventory.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/bminventory/inventory.go:6794]: :man_shrugging: sounds good

File: internal/cluster/cluster.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/cluster/cluster.go:494]: I guess I'm not sure how this avoids the issue.
In `prepareDB` we're doing this:

```
dbWithCondition = dbWithCondition.Where("status NOT IN (?)", noNeedToMonitorInStates)
return dbWithCondition
```

But then down [here](https://github.com/openshift/assisted-service/pull/2908/files/e9df23107a7219096ff7ca07f18a3491235375aa#diff-ae8489a6f9dce351a6ca906a090da834a0dba24cbb9a61221c05b6ac9b257935R116) you do:
```
err = t.prepareDB(t.db).Where("id in (?)", t.ids[t.offset:nextOffset]).Find(&clusters).Error
```
which chains another `Where` call onto the same db instance. Isn't this what you were trying to avoid?
I may not fully understand the issue though, which is why I asked.

File: internal/cluster/common.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/cluster/common.go:117]: I don't understand. Why keep an unused parameter around?

File: internal/cluster/common.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/cluster/common.go:140]: :man_shrugging: seems strange to run another database query just for an error log.

Also in the new version it seems odd to report that the error is "nothing has changed" when we failed to query the cluster at all. Wouldn't that mean we're trying to update a cluster that doesn't exist?

Maybe just log the status we used in the update query?

File: internal/host/hostcommands/install_cmd_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/host/hostcommands/install_cmd_test.go:463]: Okay, but gorm v2 didn't cause that, right?

File: internal/migrations/migrations.go (hunk 9 above)
Comment (carbonin) [reply internal/migrations/migrations.go:17]: This is important, can you put this in the PR description and commit message?

Comment (carbonin) [reply internal/migrations/migrations.go:17]: Also I guess this is the only table we have with an integer pk column?

Comment (carbonin) [reply internal/migrations/migrations.go:17]: Saw that it was added :+1:

File: internal/host/hostutil/update_host.go [hunk 21]
```diff
@@ -103,7 +103,13 @@ func UpdateHost(_ logrus.FieldLogger, db *gorm.DB, infraEnvId strfmt.UUID, hostI
...
+				srcStatus)
+		}
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/host/hostutil/update_host.go:112]: Will revert

File: internal/migrations/20210713123129_populate_infra_env_test.go (hunk 8 above)
Comment (ori-amizur) [reply internal/migrations/20210713123129_populate_infra_env_test.go:53]: Will revert

File: internal/cluster/cluster.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/cluster/cluster.go:494]: The where condition cannot be used more than once, so every time the function is called it is created again.
The previous version had a pre-created where condition that was reused many times.

File: internal/bminventory/inventory.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:1712]: OK.  We can add it to setBootstrapHost.

File: internal/cluster/cluster.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (carbonin) [reply internal/cluster/cluster.go:494]: Ah I understand now.
It continues to build up the scope even after the query is executed.

This test run from the issue you created cleared up what was going on https://github.com/go-gorm/playground/runs/4178211401?check_suite_focus=true

[the remark this answers is not in the document]
Comment (carbonin) [reply internal/cluster/cluster.go:494]: And this avoids the issue by applying the scope each time we need it rather than once at the start. Got it. Thanks @ori-amizur

File: internal/host/hostcommands/install_cmd_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/host/hostcommands/install_cmd_test.go:463]: gorm v1 ignored it.  gorm v2 crashes on it

File: internal/cluster/common.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/cluster/common.go]: Will revert.  If we want we can put in a different PR.

File: internal/bminventory/inventory.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ori-amizur) [reply internal/bminventory/inventory.go:1712]: Added to setBootstrapHost in the call to GetKnownMastersNodesIds

**Review by filanov — APPROVED** [review]

**Comment by ori-amizur** [comment]
Made upgrade tests for service and operator.  Tests were successful.
Ran the swarm test with and without this PR.  The test did not complete because of some issues with the swarm test tool.
Seems that both implementations (with and without GORM v2), have similar performance in terms of CPU and memory consumption.
